onnxruntime/core/providers/xnnpack/nn/matmul.cc

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "matmul.h"
+#include "core/graph/constants.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+#include "core/framework/transpose_helper.h"
+#include "core/providers/utils.h"
+#include "core/providers/xnnpack/detail/utils.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/providers/cpu/math/matmul_helper.h"
+
+namespace onnxruntime {
+namespace xnnpack {
+
+bool MatMul::IsMatMulOnnxNodeSupported(const NodeUnit& node_unit, const GraphViewer& graph) {
+
+  bool supported = false;
+  const onnxruntime::Node& node = node_unit.GetNode();
+
+  // use do {} while(false) so it's easier to set a breakpoint on the return
+  do {
+    auto input_defs = node.InputDefs();
+
+    if (input_defs.size() != 2) {
+      break;
+    }
+
+    const auto& A_arg = *input_defs[0];
+    const auto& B_arg = *input_defs[1];
+
+    // Support only float
+    const auto* A_type = A_arg.TypeAsProto();
+    const auto* B_type = B_arg.TypeAsProto();
+
+    if (A_type == nullptr || B_type == nullptr ||
+        A_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ||
+        B_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ) {
+        break;
+    }
+
+    // B matrix must be constant
+    if (B_arg.Exists() && graph.GetConstantInitializer(B_arg.Name(), true) == nullptr) {
+        break;
+    }
+
+    // making sure we are dealing with MatMul
+    const auto* B_shape = B_arg.Shape();
+
+    if (!B_shape || B_shape->dim_size() > 3) {
+        break;
+    }
+
+    supported = false;
+
+  } while (false);
+
+  return supported;
+}
+
+MatMul::MatMul(const OpKernelInfo& info) : OpKernel(info){
+  info.GetAttrOrDefault<int64_t>("transA", &trans_a_attr_, 0);
+  info.GetAttrOrDefault<int64_t>("transB", &trans_b_attr_, 0);
+  info.GetAttrOrDefault<float>("alpha", &alpha_attr_, 1.0);
+  int64_t trans_batch_a_attr, trans_batch_b_attr;
+  info.GetAttrOrDefault<int64_t>("transBatchA", &trans_batch_a_attr, 0);
+  info.GetAttrOrDefault<int64_t>("transBatchB", &trans_batch_b_attr, 0);
+  trans_batch_a_ = trans_batch_a_attr != 0;
+  trans_batch_b_ = trans_batch_b_attr != 0;
+}
+
+Status MatMul::PrePack(const Tensor& tensor,int input_idx, AllocatorPtr alloc,
```

Comment:
I called prepack since this was the convention - in pre-pack you call the set up xnnpack op - this in turn packs the B matrix into the op memory (so I guess it's kinda pre-packing). Do you have another flow in mind ? this is similar to the flow we have for Conv and Gemm.

Reply:
I believe you know this now, but for the benefit of anyone else reading this PR, in those cases we delayed creating the xnnpack kernel as it required the pre-packed data.